### PR TITLE
Add configurable retry mechanism for Secrets Provider deployment

### PR DESCRIPTION
**Pull request by sigalsax**
### What does this PR do?
- Removes exponential backoff and replaces with constant backoff with defaults
- Update Helm chart and JSON schema, adding RETRY_COUNT and SYNC_INTERVAL_TIME variables for a configurable retry

### What ticket does this PR close?
Connected to #174 174

#### Change log
- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

### CODE REVIEW

File: cmd/secrets-provider/main.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ func main() {
 	}
 
 	// verify we don't run with a K8s secrets storage configuration in a side car
-	if secretsConfig.StoreType == secretsConfigProvider.K8S && authnConfig.ContainerMode != "init" {
+	if secretsConfig.StoreType == secretsConfigProvider.K8S && (authnConfig.ContainerMode != "init" && authnConfig.ContainerMode != "application") {
```
Comment (orenbm) [inline cmd/secrets-provider/main.go:38]: i think you need to rebase on master. please do so and ping me when the PR is ready.

Comment (sigalsax) [reply cmd/secrets-provider/main.go]: @orenbm we should be good to go

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Red Hat certified version of the secrets-provider-for-k8s image ([cyberark/secrets-provider-for-k8s#93](https://github.com/cyberark/secrets-provider-for-k8s/pull/93))
 
 ### Changed
+- The retry backoff for the Secrets Provider is now constant instead of exponential ([cyberark/secrets-provider-for-k8s#182](https://github.com/cyberark/secrets-provider-for-k8s/pull/182))
```
Comment (orenbm) [inline CHANGELOG.md:13]: do we have an issue for this? it is best to link to issues in the CHANGELOG rather than PRs.

File: cmd/secrets-provider/main.go [hunk 3]
```diff
@@ -51,51 +52,55 @@ func main() {
...
 	}
 
-	// Configure exponential backoff
-	expBackoff := backoff.NewExponentialBackOff()
-	expBackoff.InitialInterval = 2 * time.Second
-	expBackoff.RandomizationFactor = 0.5
-	expBackoff.Multiplier = 2
-	expBackoff.MaxInterval = 15 * time.Second
-	expBackoff.MaxElapsedTime = 2 * time.Minute
+	syncIntervalTime := 3
```
Comment (orenbm) [inline cmd/secrets-provider/main.go:55]: do we want to have this as some costant?

File: cmd/secrets-provider/main.go [hunk 4]
```diff
@@ -51,51 +52,55 @@ func main() {
...
+	if envSyncIntervalVal := os.Getenv("SYNC_INTERVAL_TIME"); envSyncIntervalVal != "3" && envSyncIntervalVal != "" {
+		syncIntervalTime, _ = strconv.Atoi(envSyncIntervalVal)
+	}
```
Comment (orenbm) [inline cmd/secrets-provider/main.go:58]: it is best to move this to the config file. that is the place where we read the env to set the config

File: cmd/secrets-provider/main.go [hunk 5]
```diff
@@ -51,51 +52,55 @@ func main() {
...
+
+	retryCount := 3
+	if envRetryCount := os.Getenv("RETRY_COUNT"); envRetryCount != "3" && envRetryCount != "" {
+		retryCount, _ = strconv.Atoi(envRetryCount)
+	}
```
Comment (orenbm) [inline cmd/secrets-provider/main.go:63]: same as above. constant + read env in other file

File: helm/secrets-provider/templates/secrets-provider.yaml [hunk 6]
```diff
@@ -29,12 +29,6 @@ spec:
         imagePullPolicy: {{ .Values.secretsProvider.imagePullPolicy }}
         name: {{ .Values.secretsProvider.name }}
         env:
-        - name: SECRETS_DESTINATION
-          value: k8s_secrets
-
-        - name: CONTAINER_MODE
-          value: application
```
Comment (orenbm) [inline helm/secrets-provider/templates/secrets-provider.yaml:36]: why was this removed?

File: pkg/log/messages/info_messages.go [hunk 7]
```diff
@@ -18,7 +18,6 @@ const CSPFK003I string = "CSPFK003I Retrieving following secrets from DAP/Conjur
 const CSPFK004I string = "CSPFK004I Creating Kubernetes client"
 const CSPFK005I string = "CSPFK005I Retrieving Kubernetes secret '%s' from namespace '%s'"
 const CSPFK006I string = "CSPFK006I Updating Kubernetes secret '%s' in namespace '%s'"
-const CSPFK007I string = "CSPFK007I Waiting for %s to re-authenticate and fetch secrets."
 const CSPFK008I string = "CSPFK008I CyberArk Secrets Provider for Kubernetes v%s starting up"
 const CSPFK009I string = "CSPFK009I DAP/Conjur Secrets updated in Kubernetes successfully"
-const CSPFK010I string = "CSPFK010I Secrets Provider for Kubernetes completed successfully"
+const CSPFK010I string = "CSPFK010I Updating Kubernetes Secrets: %d retries out of %d"
```
Comment (orenbm) [inline pkg/log/messages/info_messages.go:23]: @shulifink please review

File: pkg/log/messages/error_messages.go [hunk 8]
```diff
@@ -54,4 +54,5 @@ const CSPFK036E string = "CSPFK036E Variable ID '%s' is not in the format '<acco
 const CSPFK037E string = "CSPFK037E Failed to parse DAP/Conjur variable IDs"
 
 // General
+const CSPFK039E string = "CSPFK039E Secrets Provider for Kubernetes failed to update Kubernetes Secrets"
```
Comment (orenbm) [inline pkg/log/messages/error_messages.go:57]: can you move it to be after 38?

File: pkg/log/messages/error_messages.go [hunk 9]
```diff
@@ -54,4 +54,5 @@ const CSPFK036E string = "CSPFK036E Variable ID '%s' is not in the format '<acco
...
 const CSPFK038E string = "CSPFK038E Retransmission backoff exhausted"
```
Comment (orenbm) [inline pkg/log/messages/error_messages.go:58]: is this still used?

File: cmd/secrets-provider/main.go (hunk 3 above)
Comment (sigalsax) [reply cmd/secrets-provider/main.go:55]: Yes, this is a default

File: helm/secrets-provider/templates/secrets-provider.yaml (hunk 6 above)
Comment (sigalsax) [reply helm/secrets-provider/templates/secrets-provider.yaml:36]: wasn't removed, it was pushed below with all the other configs

File: cmd/secrets-provider/main.go (hunk 5 above)
Comment (sigalsax) [reply cmd/secrets-provider/main.go:63]: they are defaults not constants

File: pkg/log/messages/error_messages.go (hunk 9 above)
Comment (sigalsax) [reply pkg/log/messages/error_messages.go:58]: yes if after 3 retries, the user should hit this log

File: cmd/secrets-provider/main.go (hunk 5 above)
Comment (orenbm) [reply cmd/secrets-provider/main.go:63]: i meant that we add a constant instead of having the `3` spreading around the code in case the default is changed. you can do this in the config file

File: pkg/log/messages/error_messages.go (hunk 9 above)
Comment (sigalsax) [reply pkg/log/messages/error_messages.go:58]: I used `printErrorAndExit` instead of log.error which is wrong. I will update this

File: cmd/secrets-provider/main.go (hunk 5 above)
Comment (sigalsax) [reply cmd/secrets-provider/main.go:63]: const list will now look like
```
const (
	K8S            = "k8s_secrets"
	CONJUR_MAP_KEY = "conjur-map"
	DEFAULT_RETRY_COUNT = 3
	DEFAULT_SYNC_INTERVAL = 3
)
```

File: pkg/log/messages/info_messages.go (hunk 7 above)
Comment (sigalsax) [reply pkg/log/messages/info_messages.go:23]: this was already reviewed

File: cmd/secrets-provider/main.go (hunk 3 above)
Comment (sigalsax) [reply cmd/secrets-provider/main.go]: added to consts

File: pkg/log/messages/error_messages.go (hunk 9 above)
Comment (sigalsax) [reply pkg/log/messages/error_messages.go:57]: jumped all the logic up to after count = retry count so the deleting of accessToken still takes place

File: cmd/secrets-provider/main.go [hunk 10]
```diff
@@ -51,63 +51,52 @@ func main() {
...
 
-	// Configure exponential backoff
-	expBackoff := backoff.NewExponentialBackOff()
-	expBackoff.InitialInterval = 2 * time.Second
-	expBackoff.RandomizationFactor = 0.5
-	expBackoff.Multiplier = 2
-	expBackoff.MaxInterval = 15 * time.Second
-	expBackoff.MaxElapsedTime = 2 * time.Minute
-
+	// Configure constant backoff for application container
```
Comment (orenbm) [inline cmd/secrets-provider/main.go:54]: what do you mean by `for application container`? this will not happen also from init container?

File: cmd/secrets-provider/main.go [hunk 11]
```diff
@@ -51,63 +51,52 @@ func main() {
...
+	constantBackOff := backoff.NewConstantBackOff(time.Duration(secretsConfig.SyncIntervalTime) * time.Minute)
+	count := 0
 	err = backoff.Retry(func() error {
```
Comment (orenbm) [inline cmd/secrets-provider/main.go:57]: i may be missing something but it seems you are not doing anything with the returned `err` object. what happens if an error is returned somewhere inside the backoff loop? will this block still run:
```
/ Deleting the retrieved Conjur access token in case we got an error after retrieval.
			// if the access token is already deleted the action should not fail
			err = accessToken.Delete()
			if err != nil {
				log.Error(messages.CSPFK003E, err)
			}
			log.Error(messages.CSPFK038E)
			printErrorAndExit(messages.CSPFK039E)
```

File: cmd/secrets-provider/main.go [hunk 12]
```diff
@@ -51,63 +51,52 @@ func main() {
...
-			if err != nil {
-				return log.RecordedError(messages.CSPFK011E)
-			}
-
-			err = provideConjurSecrets(accessToken)
-			if err != nil {
-				return log.RecordedError(messages.CSPFK016E)
-			}
-
+		if count > secretsConfig.RetryCount {
```
Comment (orenbm) [inline cmd/secrets-provider/main.go:58]: i have a suggestion that i think will make the code clearer. currently we return different errors in case they occurred and catch them after the backoff loop. Then, we clean up the token and print the errors.

However, not you add a new behaviour where you handle the error (retry count limit reached) inside the loop itself.

What you can do is return an error in this case and still catch all the errors outside of the loop.

File: helm/secrets-provider/templates/secrets-provider.yaml [hunk 13]
```diff
@@ -65,9 +59,25 @@ spec:
...
         - name: K8S_SECRETS
           value: {{ .Values.environment.k8sSecrets | join "," }}
 
+        {{- if .Values.environment.conjur.syncIntervalTime }}
+        - name: SYNC_INTERVAL_TIME
+          value: {{- .Values.environment.conjur.syncIntervalTime }}
+        {{- end }}
+
+        {{- if .Values.environment.conjur.retryCount }}
+        - name: RETRY_COUNT
```
Comment (orenbm) [inline helm/secrets-provider/templates/secrets-provider.yaml:77]: how about `RETRY_COUNT_LIMIT`? or something like that. `count` sounds like an ongoing procedure

File: cmd/secrets-provider/main.go (hunk 10 above)
Comment (sigalsax) [reply cmd/secrets-provider/main.go:54]: small left over. removing

File: pkg/secrets/config/config.go [hunk 14]
```diff
@@ -49,10 +54,22 @@ func NewFromEnv() (*Config, error) {
 		return nil, err
 	}
 
+	syncIntervalTime := DEFAULT_RETRY_COUNT
+	if envSyncIntervalVal := os.Getenv("SYNC_INTERVAL_TIME"); envSyncIntervalVal != "3" && envSyncIntervalVal != "" {
```
Comment (orenbm) [inline pkg/secrets/config/config.go:58]: you can use `DEFAULT_RETRY_COUNT`

File: cmd/secrets-provider/main.go (hunk 11 above)
Comment (sigalsax) [reply cmd/secrets-provider/main.go]: I just removed the `err =` 
if there is an error in the backoff loop, it will try again and again until that if is true
```
if count > secretsConfig.RetryCount {
}
```
Once the above is true we will process will exit with 1 error code and the AccessToken will be deleted and "Secrets Provider for Kubernetes failed to update Kubernetes Secrets" will be added to logs

Some background, I needed a way to break out of the loop because and that above if statement does just that

File: cmd/secrets-provider/main.go [hunk 15]
```diff
@@ -51,63 +51,51 @@ func main() {
...
+			log.Info(fmt.Sprintf(messages.CSPFK010I, count, secretsConfig.RetryCount))
+		}
+		count++
 
-			// Reset exponential backoff
-			expBackoff.Reset()
+		log.Info(fmt.Sprintf(messages.CSPFK001I, authn.Config.Username))
+		authnResp, err := authn.Authenticate()
+		if err != nil {
+			return log.RecordedError(messages.CSPFK010E)
```
Comment (orenbm) [inline cmd/secrets-provider/main.go:75]: i may be missing something but it seems that you are returning this error but it is never used. am i wrong?

Comment (abrahamko) [reply cmd/secrets-provider/main.go:66]: Returning this error triggers the retry loop.

File: helm/secrets-provider/values.schema.json [hunk 16]
```diff
@@ -112,6 +112,16 @@
...
               "minLength": 1
             },
+            "timeout": {
+              "type": "number",
+              "minLength": 1,
+              "minimum": 5
+            },
+            "retryCountLimit": {
+              "type": "number",
+              "minLength": 1,
```
Comment (abrahamko) [inline helm/secrets-provider/values.schema.json:122]: remove, only valid for strings and arrays.

File: helm/secrets-provider/values.schema.json [hunk 17]
```diff
@@ -112,6 +112,16 @@
               "type": "string",
               "minLength": 1
             },
+            "timeout": {
+              "type": "number",
+              "minLength": 1,
```
Comment (abrahamko) [inline helm/secrets-provider/values.schema.json:117]: remove, only valid for strings and arrays.

File: helm/secrets-provider/values.schema.json [hunk 18]
```diff
@@ -112,6 +112,16 @@
...
+              "minimum": 5
+            },
+            "retryCountLimit": {
+              "type": "number",
+              "minLength": 1,
+              "minimum": 1
```
Comment (abrahamko) [inline helm/secrets-provider/values.schema.json:123]: Why 0 is not a valid retry limit?
I may want to not retry at all.

File: helm/secrets-provider/values.schema.json [hunk 19]
```diff
@@ -112,6 +112,16 @@
               "type": "string",
               "minLength": 1
             },
+            "timeout": {
+              "type": "number",
+              "minLength": 1,
+              "minimum": 5
```
Comment (abrahamko) [inline helm/secrets-provider/values.schema.json:118]: 5 what? minutes? seconds? milliseconds?
Better call it `timeoutSec` to ensure readability.

Comment (abrahamko) [inline helm/secrets-provider/values.schema.json:118]: why minimum 5?
I would assume non-negative is enough, although in many places `timeout=-1` means forever, which is also valid.
What say you?
@oburstein-hub do you have something to add? should we allow 0 or even -1 as valid timeout values?

File: pkg/secrets/config/config.go [hunk 20]
```diff
@@ -49,10 +54,22 @@ func NewFromEnv() (*Config, error) {
...
 
+	timeout := DEFAULT_TIMEOUT
+	if envTimeout, _ := strconv.Atoi(os.Getenv("TIMEOUT")); envTimeout != DEFAULT_TIMEOUT && envTimeout != 0 {
+		timeout = envTimeout
+	}
+
+	retryCountLimit := DEFAULT_RETRY_COUNT_LIMIT
+	if envRetryCountLimit, _ := strconv.Atoi(os.Getenv("RETRY_COUNT_LIMIT")); envRetryCountLimit != DEFAULT_RETRY_COUNT_LIMIT && envRetryCountLimit != 0 {
+		retryCountLimit = envRetryCountLimit
+	}
```
Comment (abrahamko) [inline pkg/secrets/config/config.go:65]: You can extract this method into something like:
```
ParseIntFromEnvOrDefault("RETRY_COUNT_LIMIT", DEFAULT_RETRY_COUNT_LIMIT)
```
Then you can reuse it for both env vars.

File: pkg/secrets/config/config.go [hunk 21]
```diff
@@ -49,10 +54,22 @@ func NewFromEnv() (*Config, error) {
 		return nil, err
 	}
 
+	timeout := DEFAULT_TIMEOUT
+	if envTimeout, _ := strconv.Atoi(os.Getenv("TIMEOUT")); envTimeout != DEFAULT_TIMEOUT && envTimeout != 0 {
+		timeout = envTimeout
+	}
+
+	retryCountLimit := DEFAULT_RETRY_COUNT_LIMIT
+	if envRetryCountLimit, _ := strconv.Atoi(os.Getenv("RETRY_COUNT_LIMIT")); envRetryCountLimit != DEFAULT_RETRY_COUNT_LIMIT && envRetryCountLimit != 0 {
```
Comment (abrahamko) [inline pkg/secrets/config/config.go:63]: why do you check for value != 0?
Is this some kind of input validation?
If so, it should be clear. Since retry = 0 might be a valid input.
If it is to avoid the default value in case of parsing error, better check for `err != nil`, then use the default.

File: helm/secrets-provider/templates/secrets-provider.yaml [hunk 22]
```diff
@@ -65,9 +59,24 @@ spec:
...
+
+        # Enables the support of multiple Kubernetes applications
+        - name: CONTAINER_MODE
+          value: application
+
         - name: K8S_SECRETS
           value: {{ .Values.environment.k8sSecrets | join "," }}
 
+        - name: TIMEOUT
+          value: {{- .Values.environment.conjur.timeout }}
```
Comment (abrahamko) [inline helm/secrets-provider/templates/secrets-provider.yaml:73]: Since they are not mandatory, and do not have default values, what will be the value of this if user does not supply it? (which he doesn't even know he can...)

File: cmd/secrets-provider/main.go [hunk 23]
```diff
@@ -51,51 +51,42 @@ func main() {
...
-
-			fmt.Println()
-			time.Sleep(authn.Config.TokenRefreshTimeout)
+		err = retryCondition(count, secretsConfig.RetryCountLimit)
+		if err != nil {
+			return err
+		}
+		count++
+
+		log.Info(fmt.Sprintf(messages.CSPFK001I, authn.Config.Username))
```
Comment (abrahamko) [inline cmd/secrets-provider/main.go:63]: Extract to function from this line to the end of the anonymous function.
Something like: `provideSecretsToTarget` or anything else you like.

File: cmd/secrets-provider/main.go [hunk 24]
```diff
@@ -51,51 +51,42 @@ func main() {
...
+		err = accessToken.Delete()
+		if err != nil {
+			return log.RecordedError(messages.CSPFK003E, err.Error())
 		}
-	}, expBackoff)
+
+		if authnConfig.ContainerMode == "init" || authnConfig.ContainerMode == "application" {
+			log.Info(messages.CSPFK009I)
+			os.Exit(0)
+		}
```
Comment (abrahamko) [inline cmd/secrets-provider/main.go:87]: Why do you need this?
the next line `return nil` will exit the process anyway.

File: helm/secrets-provider/values.schema.json (hunk 18 above)
Comment (sigalsax) [reply helm/secrets-provider/values.schema.json:123]: @oburstein-hub ^

File: helm/secrets-provider/values.schema.json (hunk 19 above)
Comment (sigalsax) [reply helm/secrets-provider/values.schema.json:118]: name changed to retryInterval because it wasn't really a timeout

File: helm/secrets-provider/templates/secrets-provider.yaml (hunk 22 above)
Comment (sigalsax) [reply helm/secrets-provider/templates/secrets-provider.yaml]: this was updated to 
```

        {{- if .Values.environment.conjur.retryInterval }}
        - name: RETRY_INTERVAL
          value: {{- .Values.environment.conjur.retryInterval }}
        {{- end }}

        {{- if .Values.environment.conjur.retryCountLimit }}
        - name: RETRY_COUNT_LIMIT
          value: {{ .Values.environment.conjur.retryCountLimit }}
        {{- end }}

```

File: pkg/secrets/config/config.go (hunk 21 above)
Comment (sigalsax) [reply pkg/secrets/config/config.go:63]: if the user didn't provide in these env vars value they would programmatically equate to 0 so this is to ensure that this doesn't happen

File: cmd/secrets-provider/main.go (hunk 1 above)
Comment (oburstein-hub) [reply cmd/secrets-provider/main.go]: Please, turn "init" and "application" to constant String parameters

File: cmd/secrets-provider/main.go (hunk 4 above)
Comment (oburstein-hub) [reply cmd/secrets-provider/main.go]: Why do you need comparison to 3 here ? Atoi will handle it anyway, right ?

File: cmd/secrets-provider/main.go (hunk 5 above)
Comment (oburstein-hub) [reply cmd/secrets-provider/main.go]: you dont need comparison with "3" here either

File: cmd/secrets-provider/main.go (hunk 10 above)
Comment (oburstein-hub) [reply cmd/secrets-provider/main.go]: why are they multiplied by 15 and by 2 ?

File: helm/secrets-provider/values.schema.json (hunk 18 above)
Comment (oburstein-hub) [reply helm/secrets-provider/values.schema.json:121]: because it will create busy waiting

File: helm/secrets-provider/templates/secrets-provider.yaml (hunk 22 above)
Comment (oburstein-hub) [reply helm/secrets-provider/templates/secrets-provider.yaml]: why are you putting these "ifs" there ? the parameters are always in Values, arent they ?

File: helm/secrets-provider/values.schema.json (hunk 19 above)
Comment (sigalsax) [reply helm/secrets-provider/values.schema.json:117]: @oburstein-hub ^

File: cmd/secrets-provider/main.go (hunk 10 above)
Comment (sigalsax) [reply cmd/secrets-provider/main.go]: the red means it was removed @oburstein-hub

File: helm/secrets-provider/templates/secrets-provider.yaml (hunk 22 above)
Comment (sigalsax) [reply helm/secrets-provider/templates/secrets-provider.yaml]: no they are optional. If the user doesn't supply them we take defaults that are defined in code

File: helm/secrets-provider/values.schema.json (hunk 19 above)
Comment (oburstein-hub) [reply helm/secrets-provider/values.schema.json:117]: it is not convinient to put timeout 0, although -1 is a common way to sign infinite timeout. 1 can be a minimal value

File: helm/secrets-provider/templates/secrets-provider.yaml (hunk 22 above)
Comment (oburstein-hub) [reply helm/secrets-provider/templates/secrets-provider.yaml]: I think the better way it always to provide these parameters from helm chart. If the user does not want to provide them he will not put them in custom-values.yaml and defaults from values.yaml will be used

File: pkg/secrets/config/config.go [hunk 25]
```diff
@@ -49,13 +54,27 @@ func NewFromEnv() (*Config, error) {
...
 		RequiredK8sSecrets: requiredK8sSecrets,
+		RetryCountLimit:    retryCountLimit,
+		RetryInterval:      retryInterval,
 		StoreType:          storeType,
 	}, nil
 }
 
+func ParseIntFromEnvOrDefault(environmentVariable string, defaultVariable int) int {
+	retry := defaultVariable
+	if envVar, _ := strconv.Atoi(os.Getenv(environmentVariable)); envVar != defaultVariable && envVar != 0 {
```
Comment (abrahamko) [inline pkg/secrets/config/config.go:72]: instead of `envVar != 0`, which is a valid value, use `err != nil`, which indicated parsing error occurred.
Also, I suggest to document that the we failed to parse the value, what do you think?

File: helm/secrets-provider/templates/secrets-provider.yaml [hunk 26]
```diff
@@ -63,11 +57,28 @@ spec:
...
+
+        # Enables the support of multiple Kubernetes applications
+        - name: CONTAINER_MODE
+          value: application
 
         - name: K8S_SECRETS
           value: {{ .Values.environment.k8sSecrets | join "," }}
 
+        {{- if .Values.environment.conjur.retryInterval }}
+        - name: RETRY_INTERVAL
```
[the remark this answers is not in the document]
Comment (sigalsax) [reply helm/secrets-provider/templates/secrets-provider.yaml:73]: adding

File: helm/secrets-provider/values.schema.json (hunk 19 above)
Comment (sigalsax) [reply helm/secrets-provider/values.schema.json:117]: adding
